README
fortune-mod README
==================

This is fortune-mod 9708.  It is basically the same as the fortune-mod
released in October 1995, with some portability improvements, bug fixes
and documentation cleanup.

The most significant fix was for the reported bug about the `-a' option
of fortune with database names that appear in both the inoffensive and
offensive directories.  This was solved by allowing the user to append
'-o' to a fortune name while `-a' is in effect to force selection of the
offensive version of a database.

Other changes/fixes: Fortune is now consistant in how it determines a
fortune's length (for -s and -l).  The -m can now be used together with
-s or -l: only fortunes which match _BOTH_ the pattern and the length
requirement will be printed.

Most of the other features over the usual BSD fortune are summarised in
Amy's README.Linux, included below, and in the accompanying manual pages
for fortune(6) and strfile(1).

The changes in fortune.c and fortune.man are copyrighted by me but
freely distributable: see the source files for details.  All other
changes are in the public domain: you may do what you like with them.

Getting this Software
---------------------

You can find out about my latest version of this package from my
Software Hacks web page:

    http://www.progsoc.uts.edu.au/~dbugger/hacks/hacks.html

From it, there are links to download the source code.

I shall also be uploading "stable" releases to the SunSITE Linux mirror
network, so check out sunsite.unc.edu or your local SunSITE mirror. If
you still can't find it, the package should be registered under the
Linux Software Map, at:

    http://www.execpc.com/lsm/

Its LSM entry should tell you where to find it.

Cheers,

  -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>  Thu, 28 Aug 1997 11:42:15 -0400

----Amy's README.Linux---
This version of fortune is a modification of the NetBSD fortune, as
tweaked by Florian La Roche (see below, and many thanks to Florian for
starting the update), and then massively hacked on by Amy Lewis.

I (Amy) hacked on this because it was broken; the BSD source itself is
broken (I looked at it).  Specifically, if you are using an old version
of fortune, then it accesses *only* the two files "fortunes" and
"fortunes-o", even though 'fortune -[ao]f' will tell you differently.
That was my original reason to start working with the code.

Bug fixes: fortune now reads the same file list that it reports with -f.
strfile now really sorts and randomizes, instead of just setting the
'sorted' and 'randomized' flags.  strfile does not lose the pointer to a
fortune that follows a null fortune.

Enhancements: fortune -f now prints percentages, whether specified on the
command line or not.  fortune -m now prints filenames to stderr; the
fortunes printed to stdout can be redirected into a file which is valid
strfile format.  fortune -l|s can be modified with -n _number_ to specify
the number of characters in a short fortune (default 160, as before).
The means of distinguishing between offensive and inoffensive fortunes is
changed: offensive fortunes are put in a separate subdirectory.  The
contents of the fortunes databases have been extensively reviewed, and
broken into smaller, more manageable [hopefully] files.
strfile is not notably enhanced, though it received the most significant
bug fixes.  unstr now accepts a command line parameter -c _char_ which
globally changes the delimiter character.  unstr now accepts an output
file as the second file parameter, and can tell if a file has a '.dat'
extension.
An example of the use of fortune-style databases for other purposes,
called randstr, has been added.  See util/README.randstr
The Makefiles have been extensively hacked upon.

Bugs: combining -a with xx% filename, when _filename_ is found in both
the offensive and the inoffensive directories, causes fortune to exit
without an error message.  I think it's confused as to which file gets
the xx%.  I should fix this.  Don't hold your breath, though.

For more information, see the files ChangeLog, Offensive, README.install,
and cookie-files in the top-level directory, and the comments in the
various *.c source files.

Amy A. Lewis	alewis@email.unc.edu	October, 1995

----Florian's README.LINUX---
I have looked at sunsite and tsx and found one very old fortune program
and one in the debian Linux distribution. But comparing that one with
the version in NetBSD-current showed me, that NetBSD-current has fixed
so many speeling-bugs that I just had to repackage everything for the
Linux community.

In the source package are all changes for Linux in the file LINUX.DIF.
(Rewriting the Makefiles and some trivial small fixes.)

I expect this "fortune.tar.gz" to show up under /pub/Linux/games.

Not only the kernel needs speeling-corrections,

Florian La Roche      florian@jurix.jura.uni-sb.de      April 1995


PS. The following is the README from the originating NetBSD fortune:
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

#	$NetBSD: README,v 1.2 1995/03/23 08:28:29 cgd Exp $
#	@(#)README	8.1 (Berkeley) 5/31/93

The potentially offensive fortunes are not installed by default on BSD
systems.  If you're absolutely, *positively*, without-a-shadow-of-a-doubt
sure that your user community wants them installed, whack the Makefile
in the subdirectory datfiles, and do "make all install".

=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
	Some years ago, my neighbor Avery said to me: "There has not been an
adequate jokebook published since "Joe_Miller", which came out in 1739 and
which, incidentally, was the most miserable no-good ... jokebook in the
history of the printed word."
	In a subsequent conversation, Avery said: "A funny story is a funny
story, no matter who is in it - whether it's about Catholics or Protestants,
Jews or Gentiles, blacks or whites, browns or yellows.  If a story is genuinely
funny it makes no difference how dirty it is.  Shout it from the rooftops.
Let the chips fall all over the prairie and let the bonehead wowsers yelp.
... on them."
	It is a nice thing to have a neighbor of Avery's grain.  He has
believed in the aforestated principles all his life.  A great many other
people nowadays are casting aside the pietistic attitude that has led them
to plug up their ears against the facts of life.  We of The Brotherhood
believe as Avery believes; we have never been intimidated by the pharisaical
meddlers who have been smelling up the American landscape since the time of
the bundling board.  Neither has any one of our members ever been called a
racist.  Still, we have been in unremitting revolt against the ignorant
propensity which ordains, in effect, that "The Green Pastures" should never
have been written; the idiot attitude which compelled Arthur Kober to abandon
his delightful Bella Gross, and Octavius Roy Cohen to quit writing about the
splendiferous Florian Slappey; the moronic frame of mind which, if carried
to its logical end, would have forbidden Ring Lardner from writing in the
language of the masses.
		-- H. Allen Smith, "Rude Jokes"

	... let us keep in mind the basic governing philosophy of The
Brotherhood, as handsomely summarized in these words: we believe in
healthy, hearty laughter -- at the expense of the whole human race, if
needs be.
	Needs be.
		-- H. Allen Smith, "Rude Jokes"


ChangeLog
(Note: this file has been re-arranged to be in reverse chronological
order, which is The Right Thing for ChangeLogs - DLC)

August 29, 1997

  Incorporated a couple of minor changes made in the (old) Debian
  fortune-mod package, including the addition of an extra data file
  called 'cookie'.  Renamed some documentation files, and included some
  install information for non-Linux users.  Added a "-v" option to
  report the program version.

  I intend to submit this distribution to SunSITE RSN.

    -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>

May, 1997

  This release fixes many of the portability problems with the
  fortune-mod program released by Amy Lewis in October, 1995.  The
  previous version had many Linux-isms in it, which left it unworkable
  on any other platform.  This version replaces most of these with more
  standard calls, making it more likely to work under other platforms.
  The Makefiles have been modified so that GNU's gcc and make are no
  longer required: any standard make and ANSI-compatible C compiler
  should work.  Sorry, pre-ANSI compilers are not supported (c'mon, this
  is the 90's, darn it!)

  This version has been tested to work on SunOS 4.1.x as well as Linux.
  All changes made to the sources were as platform-independant as
  possible.  Therefore, no "#ifdef LINUX" or "#ifdef SUNOS4" directives
  appear in the code.  An effect of this is that a number of 'implicit
  declaration' warnings are emitted by gcc under SunOS 4.1.x, but this a
  problem with SunOS's standard headers, not with the program or the
  compiler.  The benefit of this approach is that it eases the work of
  expanding the port to include other platforms. Bug reports and fixes
  for other platforms are most welcome!

  A few "standard" C function calls were replaced with more standard
  counterparts at various points.  Generally, when there was a choice
  between a BSD version of a function and a POSIX version, the POSIX
  version was favoured (even though fortune originated on BSD).  An
  exception to this was the regex functions: either POSIX or BSD
  versions can be used, with selection made via the top-level Makefile.

  Fortune and strfile also compiles on Solaris 2.5, but a discrepency
  between the declaration and implementation of 'struct dirent' on the
  test platform caused fortune to execute incorrectly there.  It is not
  certain whether this is a bug in the header file, the C library, or
  the test platform.

  While the making of this release was not meant to become a bug
  search-and-destroy mission, some bugs were inadvertantly discovered
  and fixed. including the known bug of using -a with a file that occurs
  in both the offensive and inoffensive directories.  Fixing this
  particular bug required a way to be able to seperately identify two
  fortune files with the same name, with one in the inoffensive
  directory, the other one in the offensive directory.  Now, such a name
  will be taken to be the inoffensive file by default.  However, you can
  now append '-o' to a fortune name, and the '-o' will be removed and
  the offensive directory will be searched.  Thus you can say (assuming
  you use the distributed datfiles):

        fortune 80% politics politics-o

  Which has an 80% change of giving an "inoffensive" political fortune,
  and a 20% change of giving an "offensive" political one.  Note that
  this makes fortune-mod backwards-compatible with BSD fortune, but only
  for users, not fortune database maintainers.

  Of course, this solution only passes the buck: with the above example
  again, if you have a 'politics-o' file in your inoffensive directory,
  you are back to square one.  OTOH, seeing that '-o' originally was
  meant for offensive fortunes, using it for inoffensive ones is simply
  asking for trouble.

  Bug fixed: Fortune's definition of a fortune length (for -s and -l)
  was inconsistant.  Unordered fortunes counted the 2 delimiting
  characters (as hinted in the man page), but sorted or randomized
  fortunes did not.  Now the delimiting characters are _NEVER_ counted,
  so you will always get the length limit you expect.

  Another bug fix: -l and -s can now work together with -m.  Previously
  -l and -s were ignored when -m was in effect.  The new behaviour helps
  me count how many long or short fortunes there are in a file.

  Ansify has been removed from the package, as well as some filter
  scripts from NetBSD that no longer appear useful.  Randstr has been
  kept, but has not been improved in any way.  It at least has a man
  page: maybe somebody will find it useful.

  Some of the documentation (including the man pages) has been improved
  and updated, and some files have been renamed so that the package
  looks less Linux-specific.

    -- Dennis L. Clark <dbugger@progsoc.uts.edu.au>

Late October 1995
  Ansify has been abandoned.  I'm going to distribute this working version
  of fortune, and then see how difficult it would be to add termcap/terminfo
  enhancements to fortune itself--I don't anticipate serious problems, but
  I'd rather go ahead and get this on the net.
  
  A last-minute change was made to the way that percentages are displayed
  with -f; it is now in the format nnn.nn%.  The reason for this is that
  with the multiplication of small files, fortune -af displayed a large
  number of "0%"s--no worse than the old version, but not helpful.

  The fortunes database was finally cleaned up, and this version is now
  being distributed (at least, I hope it is).  I don't consider the current
  division of fortunes among files absolutely canonical; some are certainly
  in the wrong places.  But things are *better*.

Amy A. Lewis	alewis@email.unc.edu

Mid-October '95
  Another utility, ansify, now compiles; it has not been tested at all, so
  it may not work even slightly.
  
  Ansify is a rather stupid program, all things considered, but the work on
  it does raise an interesting possibility for an enhanced fortune.  At

util/unstr.c
/*      $NetBSD: unstr.c,v 1.3 1995/03/23 08:29:00 cgd Exp $    */

/*-
 * Copyright (c) 1991, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ken Arnold.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if 0
#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1991, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";

#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)unstr.c	8.1 (Berkeley) 5/31/93";

#endif /* not lint */
#endif /* comment out the dreck, kill the warnings */

/*
 *    This program un-does what "strfile" makes, thereby obtaining the
 * original file again.  This can be invoked with the name of the output
 * file, the input file, or both. If invoked with only a single argument
 * ending in ".dat", it is pressumed to be the input file and the output
 * file will be the same stripped of the ".dat".  If the single argument
 * doesn't end in ".dat", then it is presumed to be the output file, and
 * the input file is that name prepended by a ".dat".  If both are given
 * they are treated literally as the input and output files.
 *
 *      Ken Arnold              Aug 13, 1978
 */

/*
 * Umm.  Well, when I got this thing, it didn't work like that.  It now
 * treats the *first* filename listed as the name of the datafile; if
 * the file happens to have an extension, that's stripped off and the
 * result is the name of the strings file.  If there is no extension, then
 * the datafile has '.dat' added, and the strings file is the filename.
 * The only problem with this is if you happen to have a strings file
 * with a dot in it--in that case, specify the dat file fully.
 * 
 * The program also now accepts an optional second filename, which is the
 * name of the output file; if not specified, it dumps to stdout.
 * 
 * It can also take one parameter, which defines a new separator character.
 * This was added chiefly in order to avoid having to run sed over a
 * strings file; unstr *can* do it easily, so it should.
 *
 * We also had to add some code to make the special cases of a null fortune
 * (two separators on successive lines, a common enough error when editing
 * a strings file) and no trailing separator be treated properly.  Unstr
 * now writes out a separator string at the end of a file; strfile does
 * not consider the null string following this to be a fortune.  Be careful
 * not to put an extra newline after the required last newline--if so, you'll
 * get a fortune that contains nothing but a newline.  Karo syrup, syrup.
 * For the gory details, and lots of cussing, see strfile.c
 */
#include	<sys/types.h>
#include	<netinet/in.h>
#include	<sys/param.h>
#include	"strfile.h"
#include	<stdio.h>
#include	<ctype.h>
#include	<string.h>
#include	<unistd.h>

#ifndef MAXPATHLEN
#define	MAXPATHLEN	1024
#endif /* MAXPATHLEN */

char *Infile,			/* name of input file */
  Datafile[MAXPATHLEN],		/* name of data file */
  Delimch,			/* delimiter character */
  Outfile[MAXPATHLEN];

char NewDelch = '\0';		/* a replacement delimiter character */

FILE *Inf, *Dataf, *Outf;

/* ARGSUSED */
void getargs(int ac, char *av[])
{
    extern int optind;
    extern char *optarg;
    char *extc;
    int ch;

    while ((ch = getopt(ac, av, "c:")) != EOF)
	switch (ch)
	  {
	  case 'c':
	      NewDelch = *optarg;
	      if (!isascii(NewDelch))
	      {
		  fprintf(stderr, "Bad delimiting characher: '\\%o'\n", NewDelch);
	      }
	      break;
	  case '?':
	  default:
	      fprintf(stderr, "Usage:\n\tunstr [-c C] datafile[.ext] [outputfile]\n");
	      exit(1);
	  }

    av += optind;

    if (*av)
    {
	Infile = *av;
	fprintf(stderr, "Input file: %s\n", Infile);
	if (!strrchr(Infile, '.'))
	{
	    strcpy(Datafile, Infile);
	    strcat(Datafile, ".dat");
	}
	else
	{
	    strcpy(Datafile, Infile);
	    extc = strrchr(Infile, '.');
	    *extc = '\0';
	}
	if (*++av)
	{
	    strcpy(Outfile, *av);
	    fprintf(stderr, "Output file: %s\n", Outfile);
	}
    }
    else
    {
	fprintf(stderr, "No input file name\n");
	fprintf(stderr, "Usage:\n\tunstr [-c C] datafile[.ext] [outputfile]\n");
	exit(1);
    }
    if (!strcmp(Infile, Outfile))
    {
	fprintf(stderr, "The input file for strings (%s) must be different from the output file (%s)\n", Infile, Outfile);
	exit(1);
    }
}

void order_unstr(tbl)
     register STRFILE *tbl;
{
    register int i;
    register unsigned char *sp;
    auto off_t pos;
    char buf[BUFSIZ];
    int printedsome;

    for (i = 0; i <= tbl->str_numstr; i++)
    {
	fread((char *) &pos, 1, sizeof pos, Dataf);
	fseek(Inf, ntohl(pos), 0);
	printedsome = 0;
	for (;;)
	{
	    sp = fgets(buf, sizeof buf, Inf);
	    if (sp == NULL || STR_ENDSTRING(sp, *tbl))
	    {
		if (sp || printedsome)
		    fprintf(Outf, "%c\n", Delimch);
		break;
	    }
	    else
	    {
		printedsome = 1;
		fputs(sp, Outf);
	    }
	}
    }
}

int main(int ac, char **av)
{
    static STRFILE tbl;		/* description table */

    getargs(ac, av);
    if ((Inf = fopen(Infile, "r")) == NULL)
    {
	perror(Infile);
	exit(1);
    }
    if ((Dataf = fopen(Datafile, "r")) == NULL)
    {
	perror(Datafile);
	exit(1);
    }
    if (*Outfile == '\0')
	Outf = stdout;
    else if ((Outf = fopen(Outfile, "w+")) == NULL)
    {
	perror(Outfile);
	exit(1);
    }
    fread((char *) &tbl, sizeof tbl, 1, Dataf);
    tbl.str_version = ntohl(tbl.str_version);
    tbl.str_numstr = ntohl(tbl.str_numstr);
    tbl.str_longlen = ntohl(tbl.str_longlen);
    tbl.str_shortlen = ntohl(tbl.str_shortlen);
    tbl.str_flags = ntohl(tbl.str_flags);
    if (!(tbl.str_flags & (STR_ORDERED | STR_RANDOM)) && (!NewDelch))
    {
	fprintf(stderr, "nothing to do -- table in file order\n");
	exit(1);
    }
    if (NewDelch)
	Delimch = NewDelch;
    else
	Delimch = tbl.str_delim;
    order_unstr(&tbl);
    fclose(Inf);
    fclose(Dataf);
    fclose(Outf);
    exit(0);
}


util/strfile.c
/*      $NetBSD: strfile.c,v 1.3 1995/03/23 08:28:47 cgd Exp $  */

/*-
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Ken Arnold.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Changes, September 1995, to make the damn thing actually sort instead
 * of just pretending.  Amy A. Lewis
 * 
 * And lots more.
 * 
 * Fixed the special cases of %^J% (an empty fortune), no 'separator' at
 * the end of the file, and a trailing newline at the end of the file, all
 * of which produced total ballsup at one point or another.
 * 
 * This included adding a routine to go back and write over the last pointer
 * written or stored, for the case of an empty fortune.
 * 
 * unstr also had to be modified (well, for *lots* of reasons, but this was
 * one) to be certain to put the delimiters in the right places.
 */

/*
 * 
 #ifndef lint
 static char copyright[] =
 "@(#) Copyright (c) 1989, 1993\n\
 The Regents of the University of California.  All rights reserved.\n";
 #endif / * not lint * /
 
 #ifndef lint
 #if 0
 static char sccsid[] = "@(#)strfile.c  8.1 (Berkeley) 5/31/93";
 #else
 static char rcsid[] = "$NetBSD: strfile.c,v 1.3 1995/03/23 08:28:47 cgd Exp $";
 #endif
 #endif / * not lint * /
 *
 *I haven't the faintest flipping idea what all that is, so kill the warnings 
 */

#include	<stdlib.h>
#include	<unistd.h>
#include	<netinet/in.h>
#include	<sys/param.h>
#include	<stdio.h>
#include	<ctype.h>
#include	<string.h>
#include	"strfile.h"

#ifndef MAXPATHLEN
#define	MAXPATHLEN	1024
#endif /* MAXPATHLEN */

/*
 *    This program takes a file composed of strings seperated by
 * lines containing only the delimiting character (the default
 * character is '%') and creates another file which consists of a table
 * describing the file (structure from "strfile.h"), a table of seek
 * pointers to the start of the strings, and the strings, each terminated
 * by a null byte.  Usage:
 *
 *      % strfile [-iorsx] [ -cC ] sourcefile [ datafile ]
 *
 *      c - Change delimiting character from '%' to 'C'
 *      s - Silent.  Give no summary of data processed at the end of
 *          the run.
 *      o - order the strings in alphabetic order
 *      i - if ordering, ignore case 
 *      r - randomize the order of the strings
 *      x - set rotated bit
 *
 *              Ken Arnold      Sept. 7, 1978 --
 *
 *      Added ordering options.
 * 
 * Made ordering options do more than set the bloody flag, September 95 A. Lewis
 * 
 * Always make sure that your loop control variables aren't set to bloody
 * *zero* before distributing the bloody code, all right?
 * 
 */

#define	TRUE	1
#define	FALSE	0

#define	STORING_PTRS	(Oflag || Rflag)
#define	CHUNKSIZE	512

#define	ALWAYS	1
#define	ALLOC(ptr,sz)	if (ALWAYS) { \
			if (ptr == NULL) \
				ptr = malloc((unsigned int) (CHUNKSIZE * sizeof *ptr)); \
			else if (((sz) + 1) % CHUNKSIZE == 0) \
				ptr = realloc((void *) ptr, ((unsigned int) ((sz) + CHUNKSIZE) * sizeof *ptr)); \
			if (ptr == NULL) { \
				fprintf(stderr, "out of space\n"); \
				exit(1); \
			} \
		}

typedef struct
{
    char first;
    off_t pos;
}
STR;

char *Infile = NULL,		/* input file name */
  Outfile[MAXPATHLEN] = "",	/* output file name */
  Delimch = '%';		/* delimiting character */

int Sflag = FALSE;		/* silent run flag */
int Oflag = FALSE;		/* ordering flag */
int Iflag = FALSE;		/* ignore case flag */
int Rflag = FALSE;		/* randomize order flag */
int Xflag = FALSE;		/* set rotated bit */
long Num_pts = 0;		/* number of pointers/strings */

off_t *Seekpts;

FILE *Sort_1, *Sort_2;		/* pointers for sorting */

STRFILE Tbl;			/* statistics table */

STR *Firstch;			/* first chars of each string */

void usage(void)
{
    fprintf(stderr,
	    "strfile [-iorsx] [-c char] sourcefile [datafile]\n");
    exit(1);
}

/*
 *    This routine evaluates arguments from the command line
 */
void getargs(int argc, char **argv)
{
    extern char *optarg;
    extern int optind;
    int ch;

    while ((ch = getopt(argc, argv, "c:iorsx")) != EOF)
	switch (ch)
	  {
	  case 'c':		/* new delimiting char */
	      Delimch = *optarg;
	      if (!isascii(Delimch))
	      {
		  printf("bad delimiting character: '\\%o\n'",
			 Delimch);
	      }
	      break;
	  case 'i':		/* ignore case in ordering */
	      Iflag++;
	      break;
	  case 'o':		/* order strings */
	      Oflag++;
	      break;
	  case 'r':		/* randomize pointers */
	      Rflag++;
	      break;
	  case 's':		/* silent */
	      Sflag++;
	      break;
	  case 'x':		/* set the rotated bit */
	      Xflag++;
	      break;
	  case '?':
	  default:
	      usage();
	  }
    argv += optind;

    if (*argv)
    {
	Infile = *argv;
	if (*++argv)
	    (void) strcpy(Outfile, *argv);
    }
    if (!Infile)
    {
	puts("No input file name");
	usage();
    }
    if (*Outfile == '\0')
    {
	strcpy(Outfile, Infile);
	strcat(Outfile, ".dat");
    }
}

/*
 * add_offset:
 *      Add an offset to the list, or write it out, as appropriate.
 */
void add_offset(FILE * fp, off_t off)
{
    off_t net;

    if (!STORING_PTRS)
    {
	net = htonl(off);
	fwrite(&net, 1, sizeof net, fp);
    }
    else
    {
	ALLOC(Seekpts, Num_pts + 1);
	Seekpts[Num_pts] = off;
    }
    Num_pts++;
}

/*
 * fix_last_offset:
 *     Used when we have two separators in a row.
 */
void fix_last_offset(FILE * fp, off_t off)
{
    off_t net;

    if (!STORING_PTRS)
    {
	net = htonl(off);
	fseek(fp, -(sizeof net), SEEK_CUR);
	fwrite(&net, 1, sizeof net, fp);
    }
    else
	Seekpts[Num_pts - 1] = off;
}

/*
 * cmp_str:
 *      Compare two strings in the file
 */
int cmp_str(const void *v1, const void *v2)
{
    register int c1, c2;
    register int n1, n2;
    register STR *p1, *p2;

#define	SET_N(nf,ch)	(nf = (ch == '\n'))
#define	IS_END(ch,nf)	(ch == Delimch && nf)

    p1 = (STR *) v1;
    p2 = (STR *) v2;
    c1 = p1->first;
    c2 = p2->first;
    if (c1 != c2)
	return c1 - c2;

    fseek(Sort_1, p1->pos, 0);
    fseek(Sort_2, p2->pos, 0);

    n1 = FALSE;
    n2 = FALSE;
    while (!isalnum(c1 = getc(Sort_1)) && c1 != '\0')
	SET_N(n1, c1);
    while (!isalnum(c2 = getc(Sort_2)) && c2 != '\0')
	SET_N(n2, c2);

    while (!IS_END(c1, n1) && !IS_END(c2, n2))
    {
	if (Iflag)
	{
	    if (isupper(c1))
		c1 = tolower(c1);
	    if (isupper(c2))
		c2 = tolower(c2);
	}
	if (c1 != c2)
	    return c1 - c2;
	SET_N(n1, c1);
	SET_N(n2, c2);
	c1 = getc(Sort_1);
	c2 = getc(Sort_2);
    }
    if (IS_END(c1, n1))
	c1 = 0;
    if (IS_END(c2, n2))
	c2 = 0;
    return c1 - c2;
}

/*
 * do_order:
 *      Order the strings alphabetically (possibly ignoring case).
 */
void
  do_order(void)
{
    register long i;
    register off_t *lp;
    register STR *fp;

    Sort_1 = fopen(Infile, "r");
    Sort_2 = fopen(Infile, "r");
    qsort((char *) Firstch, (int) Num_pts - 1, sizeof *Firstch, cmp_str);
/*      i = Tbl.str_numstr;
 * Fucking brilliant.  Tbl.str_numstr was initialized to zero, and is still zero
 */
    i = Num_pts - 1;
    lp = Seekpts;
    fp = Firstch;
    while (i--)
	*lp++ = fp++->pos;
    fclose(Sort_1);
    fclose(Sort_2);
    Tbl.str_flags |= STR_ORDERED;
}

char *
  unctrl(char c)
{
    static char buf[3];

    if (isprint(c))
    {
	buf[0] = c;
	buf[1] = '\0';
    }
    else if (c == 0177)
    {
	buf[0] = '^';
	buf[1] = '?';
    }
    else
    {
	buf[0] = '^';
	buf[1] = c + 'A' - 1;
    }
    return buf;
}

/*
 * randomize:
 *      Randomize the order of the string table.  We must be careful
 *      not to randomize across delimiter boundaries.  All
 *      randomization is done within each block.
 */
void randomize(void)
{
    register int cnt, i;
    register off_t tmp;
    register off_t *sp;
    extern time_t time(time_t *);

    srandom((int) (time((time_t *) NULL) + getpid()));

    Tbl.str_flags |= STR_RANDOM;
/*      cnt = Tbl.str_numstr;
 * See comment above.  Isn't this stuff distributed worldwide?  How embarrassing!
 */
    cnt = Num_pts;

    /*
     * move things around randomly
     */

    for (sp = Seekpts; cnt > 0; cnt--, sp++)
    {
	i = random() % cnt;
	tmp = sp[0];
	sp[0] = sp[i];
	sp[i] = tmp;
    }
}

/*
 * main:
 *      Drive the sucker.  There are two main modes -- either we store
 *      the seek pointers, if the table is to be sorted or randomized,
 *      or we write the pointer directly to the file, if we are to stay
 *      in file order.  If the former, we allocate and re-allocate in
 *      CHUNKSIZE blocks; if the latter, we just write each pointer,
 *      and then seek back to the beginning to write in the table.
 */
int main(int ac, char **av)
{
    register unsigned char *sp;
    register FILE *inf, *outf;
    register off_t last_off, length, pos, *p;
    register int first, cnt;
    register char *nsp;
    register STR *fp;
    static char string[257];

    getargs(ac, av);		/* evalute arguments */
    if ((inf = fopen(Infile, "r")) == NULL)
    {
	perror(Infile);
	exit(1);
    }

    if ((outf = fopen(Outfile, "w")) == NULL)
    {
	perror(Outfile);
	exit(1);
    }
    if (!STORING_PTRS)
	(void) fseek(outf, sizeof Tbl, 0);

    /*
     * Write the strings onto the file
     */

    Tbl.str_longlen = 0;
    Tbl.str_shortlen = (unsigned int) 0xffffffff;
    Tbl.str_delim = Delimch;
    Tbl.str_version = VERSION;
    first = Oflag;
    add_offset(outf, ftell(inf));
    last_off = 0;
    do
    {
	sp = fgets(string, 256, inf);
	if (sp == NULL || STR_ENDSTRING(sp, Tbl))
	{
	    pos = ftell(inf);
	    length = pos - last_off - (sp ? strlen(sp) : 0);
	    if (!length)
		/* Here's where we go back and fix things, if the
		 * 'fortune' just read was the null string.
		 * We had to make the assignment of last_off slightly
		 * redundant to achieve this.
		 */
	    {
		if (pos - last_off == 2)
		    fix_last_offset(outf, pos);
		last_off = pos;
		continue;
	    }
	    last_off = pos;
	    add_offset(outf, pos);
	    if (Tbl.str_longlen < length)
		Tbl.str_longlen = length;
	    if (Tbl.str_shortlen > length)
		Tbl.str_shortlen = length;
	    first = Oflag;
	}
	else if (first)
	{
	    for (nsp = sp; !isalnum(*nsp); nsp++)
		continue;
	    ALLOC(Firstch, Num_pts);
	    fp = &Firstch[Num_pts - 1];
	    if (Iflag && isupper(*nsp))
		fp->first = tolower(*nsp);
	    else
		fp->first = *nsp;
	    fp->pos = Seekpts[Num_pts - 1];
	    first = FALSE;
	}
    }
    while (sp != NULL);

    /*
     * write the tables in
     */

    fclose(inf);

    if (Oflag)
	do_order();
    else if (Rflag)
	randomize();

    if (Xflag)
	Tbl.str_flags |= STR_ROTATED;

    if (!Sflag)
    {
	printf("\"%s\" created\n", Outfile);
	if (Num_pts == 2)
	    puts("There was 1 string");
	else
	    printf("There were %ld strings\n", Num_pts - 1);
	printf("Longest string: %lu byte%s\n", Tbl.str_longlen,
	       Tbl.str_longlen == 1 ? "" : "s");
	printf("Shortest string: %lu byte%s\n", Tbl.str_shortlen,
	       Tbl.str_shortlen == 1 ? "" : "s");
    }

    fseek(outf, (off_t) 0, 0);
    Tbl.str_version = htonl(Tbl.str_version);
    Tbl.str_numstr = htonl(Num_pts - 1);
    /* Look, Ma!  After using the variable three times, let's store
     * something in it!
     */
    Tbl.str_longlen = htonl(Tbl.str_longlen);
    Tbl.str_shortlen = htonl(Tbl.str_shortlen);
    Tbl.str_flags = htonl(Tbl.str_flags);
    fwrite((char *) &Tbl, sizeof Tbl, 1, outf);
    if (STORING_PTRS)
    {
	for (p = Seekpts, cnt = Num_pts; cnt--; ++p)
	    *p = htonl(*p);
	fwrite((char *) Seekpts, sizeof *Seekpts, (int) Num_pts, outf);
    }
    fclose(outf);
    exit(0);
}
